torch/serialization.py

```diff
@@ -207,6 +207,46 @@ def __exit__(self, *args):
             self.file_like.close()
 
 
+class _open_zipfile_writer(_open_file_like):
```

Comment:
There are 4 classes `{buffer, file} x {input, output}` hidden in these two class, making the implementation have excessive if statements. It would make more sense to make this _open_file_like interface instanciate the right class on construction rather than defer it to later actions.

Reply:
Completely separating these apart will result in something like 8 classes (4 for zip files, 2 for regular files, plus probably some base classes) which seems way more complex than the current implementation.

Reply:
I don't think this needs to be 8 separate classes, but the current way it is organized is confusing because every function starts with an if statement. It would be a good idea to ask what the responsibility of a _open_file_like object actually is, and build object oriented abstractions for it.